lua/wire/zvm/zvm_tests.lua

```diff
@@ -0,0 +1,411 @@
+			for _,i in ipairs(files) do
+				ZVMTestSuite.TestFiles[#ZVMTestSuite.TestFiles+1] = i
+			end
+	end
+	if #ZVMTestSuite.TestFiles == 0 then
+		if names ~= nil then
+			if names ~= "" then
+				print("Didn't find any tests with name(s): "..names)
+				return
+			end
+		end
+		ZVMTestSuite.RunAll()
+	else
+		PrintTable(ZVMTestSuite.TestFiles)
+		ZVMTestSuite.StartTesting()
+	end
+end
+
+function ZVMTestSuite.RunAll()
+	local files,directories = file.Find(testDirectory..'/*.lua',"LUA","nameasc")
+	ZVMTestSuite.TestFiles = files or {}
+	ZVMTestSuite.StartTesting()
+end
+
+function ZVMTestSuite.StartTesting()
+	ZVMTestSuite.TestQueue = {}
+	ZVMTestSuite.TestStatuses = {}
+	for ind,i in ipairs(ZVMTestSuite.TestFiles) do -- copy with reversed indexes so we can use cheap popping
+		ZVMTestSuite.TestQueue[(#ZVMTestSuite.TestFiles)+1-ind] = i
+	end
+	print(#ZVMTestSuite.TestFiles.." tests loaded")
+	ZVMTestSuite.RunNextTest()
+end
+
+function ZVMTestSuite.FinishTest(fail)
+	local finalFail = false
+	if fail == nil then
+		finalFail = true
+	else
+		finalFail = fail
+	end
+	if ZVMTestSuite.CurrentWarnings > 0 then
+		print("Compiler Warnings from "..ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue]..": "..ZVMTestSuite.CurrentWarnings)
+		ZVMTestSuite.CurrentWarnings = 0
+	end
+	ZVMTestSuite.TestStatuses[#ZVMTestSuite.TestStatuses+1] = finalFail -- auto fail on return nil
+	ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue] = nil
+	if #ZVMTestSuite.TestQueue > 0 then
+		ZVMTestSuite.RunNextTest()
+	else
+		local passed, failed = 0, 0
+		for ind,i in ipairs(ZVMTestSuite.TestFiles) do
+			if ZVMTestSuite.TestStatuses[ind] then
+				failed = failed + 1
+				MsgC(Color(255,0,0),"Error ",Color(255,255,255),"in "..i.."\n")
+			else
+				passed = passed + 1
+				--MsgC(Color(0,255,0),i.." passed tests".."\n")
+			end
+		end
+		local passmod, errormod = "",""
+		if passed > 1 then
+			passmod = "s"
+		end
+		if failed > 1 then
+			errormod = "s"
+		end
+		print(failed.." Failed test"..errormod..", "..passed.." Passed test"..passmod..", Compiler Warnings: "..ZVMTestSuite.Warnings)
+	end
+end
+
+function ZVMTestSuite.Error(...)
+	local args = table.Pack(...)
+	MsgC(Color(255,0,0),"in file ",Color(255,255,255),ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue],Color(255,0,0)," Error: ")
+	if args ~= nil then
+		if istable(args) then
+			for _,i in pairs(args) do
+				MsgC(Color(255,255,255),i)
+			end
+		else
+			MsgC(Color(255,255,255),tostring(args))
+		end
+	end
+	MsgC(Color(0,0,255),'\n')
+end
+
+function ZVMTestSuite.RunNextTest()
+	local curVM = CPULib.VirtualMachine()
+	ZVMTestSuite.Initialize(curVM)
+	print("Running "..ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue])
+	include(testDirectory..'/'..ZVMTestSuite.TestQueue[#ZVMTestSuite.TestQueue])
+	CPUTest:RunTest(curVM,ZVMTestSuite)
+end
+
+function ZVMTestSuite:LoadFile(FileName)
+	return file.Read('lua/'..testDirectory..'/'..FileName,"GAME")
```

Comment:
I feel like tests should include the source they're testing inside rather than be split up like this

Reply:
This is primarily for cases that require the compiler to grab the files itself, the includes.lua example definitely needs this, I agree with you to the degree that they _should_ include most of the code now that I know about multi-line strings being possible with [[]].

Though, I could also make this into a "virtual file system" by having this fetch strings from a table named something like "files" provided by the cputest, if that sounds more appealing than leaving the txt files in the tests folder.